Add option to disable X11 linking

Technius opened the pull request:
This adds a CMake option to disable linking with X11, enabling builds on some platforms (e.g., Mac M1 builds using the Nix package manager).

Febbe commented:
This is already implemented via: [`CMAKE_DISABLE_FIND_PACKAGE_<PackageName>`](https://cmake.org/cmake/help/latest/variable/CMAKE_DISABLE_FIND_PACKAGE_PackageName.html) and [CMAKE_REQUIRE_FIND_PACKAGE_<PackageName>](https://cmake.org/cmake/help/latest/variable/CMAKE_REQUIRE_FIND_PACKAGE_PackageName.html#variable:CMAKE_REQUIRE_FIND_PACKAGE_%3CPackageName%3E)

Also, it does not feel good to add so much boilerplate code.

Febbe commented on CMakeLists.txt:
```diff
@@ -140,6 +139,35 @@ if (GtkSourceView_FOUND)
 endif ()
 
+# X11 link settings
+option(FORCE_X11 "Force linking with X11" OFF)
+if (NOT FORCE_X11 AND (CMAKE_SYSTEM_NAME MATCHES "Windows|Darwin"))
+  set(_ENABLE_X11_default OFF)
+else()
+  set(_ENABLE_X11_default ON)
+endif()
+option(ENABLE_X11 "Link with X11" ${_ENABLE_X11_default})
+unset(_ENABLE_X11_default)
+
+if (ENABLE_X11)
+    message(STATUS "Link with X11 enabled, checking for X11...")
+    if (FORCE_X11)
+        find_package(X11 REQUIRED)
+    else()
+        find_package(X11)
+    endif()
+
+    if (X11_FOUND)
+        message(STATUS "Found X11, checking required dependencies...")
+        if (NOT X11_Xi_FOUND)
+            message(SEND_ERROR "X11 support enabled, but required library libXi not found")
+        endif ()
+        if (NOT X11_Xext_FOUND)
+            message(SEND_ERROR "X11 support enabled, but required library libXext not found")
+        endif ()
+    endif ()
+endif ()
+
```
I would modify the linking, so that the generator expression awaits both libraries, if X11 has been found.

```suggestion
# The following variables are not required, just to show the user that they are there.
set(CMAKE_DISABLE_FIND_PACKAGE_X11 OFF CACHE BOOL "Disable X11")
set(CMAKE_REQUIRE_FIND_PACKAGE_X11 OFF CACHE BOOL "Force X11")
find_package(X11 #[[COMPONENTS Xi Xext X11]]) # component might not work, and must be tested (not documented).
```

Technius commented:
Thanks for the review.

* I wasn't aware of `CMAKE_DISABLE/REQUIRE_FIND_PACKAGE_*`; thanks for pointing it out. Modified the build to use those two files.
* Added Xi and Xext as components (it didn't seem very obvious from the FindX11.cmake script that they are components). Tested working with my local build.

Febbe commented:
> Added Xi and Xext as components (it didn't seem very obvious from the FindX11.cmake script that they are components). Tested working with my local build.

Yes, I also didn't know if they work.

Technius commented:
Merging in 24 hours unless there are objections.
